src/proxy.c

```diff
@@ -277,12 +273,26 @@ cc_proxy_msg_read (GIOChannel *source, GIOCondition condition,
 		}
 	}
 
-	/* ensure DO NOT copy extra bytes */
-	bytes_to_copy = MIN((size_t)bytes_read-data_offset, (size_t)p_msg.length);
-	g_string_append_len(proxy_data->msg_received,
-		iov_buffer+data_offset, (ssize_t)bytes_to_copy);
-
-	if (bytes_read >= p_msg.length+data_offset) {
+	/* cc-proxy sends a fd as out-of-band data
+	 * followed by an 'F'.
+	 * *DO NOT* copy these characters to msg_received
+	 */
+	if (bytes_read > 1 || iov_buffer[0] != 'F') {
+		memcpy (buf_len, iov_buffer, sizeof (buf_len));
+		p_msg.length = cc_oci_get_big_endian_32 (buf_len);
+		g_debug ("msg length: %d", p_msg.length);
+
+		/* ensure DO NOT copy extra bytes */
+		bytes_to_copy = MIN ((size_t)bytes_read-data_offset, (size_t)p_msg.length);
+		g_string_append_len (proxy_data->msg_received,
+			iov_buffer+data_offset, (ssize_t)bytes_to_copy);
+
+		/* all message was read */
+		if (bytes_read >= p_msg.length+data_offset) {
+			goto out;
+		}
+	} else {
+		/* only out-of-band data was received */
 		goto out;
```

Comment:
Is this `goto` safe? It jumps down to a `g_debug()` call that displays the received message... but there isn't one in the OOB message scenario right? Do we need a new `out2` label maybe?